crates/sui-rpc-api/proto/sui/rpc/v2alpha/move_descriptor.proto

```diff
@@ -0,0 +1,210 @@
+  KEY = 4;
+}
+
+// A generic type parameter used in the declaration of a struct or enum.
+message TypeParameter {
+  // The type parameter constraints
+  repeated Ability constraints = 1;
+
+  // Whether the parameter is declared as phantom
+  optional bool is_phantom = 2;
+}
+
+// Descriptor of a field that belongs to a struct or enum variant
+message FieldDescriptor {
+  // Name of the field
+  optional string name = 1;
+
+  // Order or position of the field in the struct or enum variant definition.
+  optional uint32 position = 2;
+
+  // The type of the field
+  optional OpenSignatureBody type = 5;
+}
+
+enum Type {
+  TYPE_UNKNOWN = 0;
+  ADDRESS = 1;
+  BOOL = 2;
+  U8 = 3;
+  U16 = 4;
+  U32 = 5;
+  U64 = 6;
+  U128 = 7;
+  U256 = 8;
+  VECTOR = 9;
+  DATA_TYPE = 10;
+  TYPE_PARAMETER = 11;
+}
+
+// Descriptor of an enum variant
+message VariantDescriptor {
+  // Name of the variant
+  optional string name = 1;
+
+  // Order or position of the variant in the enum definition.
+  optional uint32 position = 2;
+
+  // Set of fields defined by this variant.
+  repeated FieldDescriptor fields = 3;
+}
+
+message OpenSignatureBody {
+  // Type of this signature
+  optional Type type = 1;
+
+  // position of the type parameter as defined in the containing data type descriptor when `type` is `TYPE_PARAMETER`
+  optional uint32 type_parameter = 2;
+
+  // Fully qualified name of the data type when `type` is `DATA_TYPE`
+  optional string type_name = 3;
+
+  // Set when `type` is `VECTOR` or `DATA_TYPE`
+  repeated OpenSignatureBody type_parameter_instantiation = 4;
+}
+
+// Descriptor of a Move function
+message FunctionDescriptor {
+  // Name of the function
+  optional string name = 1;
+
+  // XXX Do we want both storage/runtime info here??
+  // PackageId where this function is defined.
+  optional string package_id = 2;
+
+  // Name of the module where this function is defined
+  optional string module = 3;
+
+  // Fully qualified name of this function.
+  //
+  // This is `<package_id>::<module>::<name>`
+  optional string function_name = 4;
+  
+  // Whether the function is `public`, `private` or `public(friend)`
+  optional Visibility visibility = 5;
+
+  // Whether the function is marked `entry` or not.
+  optional bool is_entry = 6;
+
+  // Ability constraints for type parameters
+  repeated Ability type_parameters = 7;
+
+  // Formal parameter types.
+  repeated OpenSignature parameters = 8;
+
+  // Return types.
+  repeated OpenSignature returns = 9;
+}
+
+enum Reference {
+  REFERENCE_UNKNOWN = 0;
+  IMMUTABLE = 1;
+  MUTABLE = 2;
+}
+
+enum Visibility {
+  VISIBILITY_UNKNOWN = 0;
+  PRIVATE = 1;
+  PUBLIC = 2;
+  FRIEND = 3;
+}
+
+message OpenSignature {
+  optional Reference reference = 1;
+  optional OpenSignatureBody body = 2;
+}
+
+message Module {
+  // Name of this module
+  optional name = 1;
+
+  // PackageId of the package where this module is defined.
+  optional package_id = 2;
```

Comment:
Just a note: the address in the module's self ID will not match the storage ID of the package outside of the first version of the package -- module's are all in terms of runtime/original ID when you load them. So just something to be aware of since I'm not sure which one you're intending for this to be.

Reply:
how does this work when it comes to calling a function from one version of a module vs another?